numpy/core/tests/test_scalarbuffer.py

```diff
@@ -0,0 +1,84 @@
+"""
+Test scalar buffer interface adheres to PEP 3118
+"""
+import sys
+import numpy as np
+from numpy.testing import run_module_suite, assert_, assert_equal, dec
+
+# PEP3118 format strings for native (standard alignment and byteorder) types
+scalars_and_codes = [
+    (np.bool_, '?'),
+    (np.byte, 'b'),
+    (np.short, 'h'),
+    (np.intc, 'i'),
+    (np.int_, 'l'),
+    (np.longlong, 'q'),
+    (np.ubyte, 'B'),
+    (np.ushort, 'H'),
+    (np.uintc, 'I'),
+    (np.uint, 'L'),
+    (np.ulonglong, 'Q'),
+    (np.half, 'e'),
+    (np.single, 'f'),
+    (np.double, 'd'),
+    (np.longdouble, 'g'),
+    (np.csingle, 'Zf'),
+    (np.cdouble, 'Zd'),
+    (np.clongdouble, 'Zg'),
+]
+
+
+class TestScalarPEP3118(object):
+    skip_if_no_buffer_interface = dec.skipif(sys.version_info.major < 3,
+                "scalars do not implement buffer interface in Python 2")
+
+    @skip_if_no_buffer_interface
+    def test_scalar_match_array(self):
+        for scalar, _ in scalars_and_codes:
+            x = scalar()
+            a = np.array([], dtype=np.dtype(scalar))
+            mv_x = memoryview(x)
+            mv_a = memoryview(a)
+            assert_equal(mv_x.format, mv_a.format)
+
+    @skip_if_no_buffer_interface
+    def test_scalar_dim(self):
+        for scalar, _ in scalars_and_codes:
+            x = scalar()
+            mv_x = memoryview(x)
+            assert_equal(mv_x.itemsize, np.dtype(scalar).itemsize)
+            assert_equal(mv_x.ndim, 0)
+            assert_equal(mv_x.shape, ())
+            assert_equal(mv_x.strides, ())
+            assert_equal(mv_x.suboffsets, ())
+
+    @skip_if_no_buffer_interface
+    def test_scalar_known_code(self):
+        for scalar, code in scalars_and_codes:
+            x = scalar()
+            mv_x = memoryview(x)
+            assert_equal(mv_x.format, code)
+
+    @skip_if_no_buffer_interface
+    def test_void_scalar_structured_data(self):
+        dt = np.dtype([('name', np.unicode_, 16), ('grades', np.float64, (2,))])
+        x = np.array(('ndarray_scalar', (1.2, 3.0)), dtype=dt)[()]
+        assert_(isinstance(x, np.void))
+        mv_x = memoryview(x)
+        expected_size = 16 * np.dtype((np.unicode_, 1)).itemsize
+        expected_size += 2 * np.dtype((np.float64, 1)).itemsize
```

Comment:
Deliberately not using `expected_size = dt.itemsize`?

Reply:
Yes. I'd like to test against an explicit values, rather than trusting that dt.itemsize is correct.
I do use dtype.itemsize of `unicode_` and `float64` when calculating `expected_size` because of platform dependant sizes of `unicode_`. But this is still one step more explicit than testing scalar void.itemsize ==  dtype void.itemsize

Reply:
The assertion memoryview of void itemsize, not scalar void itemsize - arguably all we care about is that the the `itemsize` of a memoryview and dtype match.

I don't feel too strongly about this though.

Reply:
> because of platform dependant sizes of unicode_

Note that `unicode` as a dtype is always 4 byte-characters - you're thinking of `unicode_` scalars, which are sometimes 2-byte, due to subclassing the builtin `unicode`.